go/tools/gopackagesdriver/aspect.bzl

```diff
@@ -49,32 +49,53 @@ def file_path(f):
         prefix = "__BAZEL_OUTPUT_BASE__"
     return paths.join(prefix, f.path)
 
-def _go_archive_to_pkg(archive):
-    go_files = [
-        file_path(src)
-        for src in archive.data.srcs
-        if src.path.endswith(".go")
-    ]
-    return struct(
-        ID = str(archive.data.label),
-        PkgPath = archive.data.importpath,
-        ExportFile = file_path(archive.data.export_file),
-        GoFiles = go_files,
-        CompiledGoFiles = go_files,
-        OtherFiles = [
-            file_path(src)
-            for src in archive.data.srcs
-            if not src.path.endswith(".go")
-        ],
-        Imports = {
-            pkg.data.importpath: str(pkg.data.label)
-            for pkg in archive.direct
-        },
-    )
+def _escape_special_chars(value, delimiter):
+    # We only have "\" and "=" as special characters.
+    # Double escape the escaper char.
+    escaped = value.replace("\\", "\\\\")
+
+    # Escape delimiter.
+    return escaped.replace(delimiter, "\\" + delimiter)
+
+def format_key_value_arg(key, value, delimiter = "="):
+    escaped_key = _escape_special_chars(key, delimiter)
+    escaped_value = _escape_special_chars(value, delimiter)
+    return "{}{}{}".format(escaped_key, delimiter, escaped_value)
 
-def make_pkg_json(ctx, name, pkg_info):
+def _go_archive_to_pkg_json(ctx, name, archive):
```

Comment:
As an optimization, let's keep the old code that builds the JSON at analysis time and only run this action if we know there's a tree artifact. Running an action is going to be slower, and this is latency-sensitive, so let's only do it when needed.